### ISSUE TITLE
autodoc.py refactoring: special_attrgetters confusion.

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#106]
While implementing a zope autointerface extension I I found the special_attrgetters behaviour confusing for the following reason:

I implemented InterfaceDocumenter and InterfaceAttributeDocumenter classes.  I assumed that only the first required special_attrgetter since one does not need to get attributes of Attributes.

However, when testing the attribute for importability, import_object() calls get_attr along the entire chain using the InterfaceAttributeDocumenter instance as self, thus, trying to access the Attribute from the Interface using InterfaceAttributeDocumenter.special_attrgetter.

It is not clear to me how to easily refactor the code so that import_object() always uses the correct special_attrgetter.

If this cannot easily been done, a strong note needs to be made about this in the documentation and/or code.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/106
- Originally reported by: Michael Forbes
- Originally created at: 2009-02-19T17:21:46.299

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2009-02-19 16:33:21+00:00_

I agree that this is a bit confusing. {{{special_attrgetter}}} is not intended to be overridden, but the one dict that is in Documenter should be added to. In your case, you'd do

{{{Documenter.special_attrgetter[Interface] = interface_getattr}}}

Maybe a setup function on the app, like {{{add_documenter}}}, is in order here.

I hope you find the rest of the API convenient to write custom documenters for? If not, please tell me, since it is not yet set in stone.

**Comment by shimizukawa** [comment]
_From Michael Forbes on 2009-02-19 17:04:19+00:00_

Here is a simple fix: Just update the {{{Documenter.special_attrgetter}}} attribute in {{{add_documenter}}}:

Patch against 9f5725038f9d (including fix for issue #107)

{{{
diff -r 9f5725038f9d sphinx/ext/autodoc.py
--- a/sphinx/ext/autodoc.py Thu Feb 19 16:15:59 2009 +0100
+++ b/sphinx/ext/autodoc.py Thu Feb 19 10:01:25 2009 -0700
@@ -454,7 +454,8 @@

```
 def get_object_members(self, want_all):
     """
```
-        Return (membername, member) pairs of the members of _self.object_.
-        Return `(members_check_module, members)` where `members` is a
- ```
     list of `(membername, member)` pairs of the members of *self.object*.
  
   If *want_all* is True, return all members.  Else, only return those
   members given by *self.options.members* (which may also be none).
  ```
  
  @@ -479,8 +480,9 @@
           # **dict** contains only the members directly defined in
           # the class (but get them via getattr anyway, to e.g. get
           # unbound method objects instead of function objects)
-            return False, sorted([(mname, self.get_attr(self.object, mname))
-                                  for mname in self.object.**dict**])
-            return False, sorted([
-                (mname, self.get_attr(self.object, mname))
- ```
             for mname in self.get_attr(self.object, '__dict__')])
  ```
  
   def filter_members(self, members, want_all):
       """
  @@ -1043,6 +1045,10 @@
   if not issubclass(cls, Documenter):
       raise ExtensionError('autodoc documenter %r must be a subclass '
                            'of Documenter' % cls)
  +
-    # Maybe check and warn if attgetters are overwritten.
-    Documenter.special_attrgetters.update(cls.special_attrgetters)
  +
   # actually, it should be possible to override Documenters
   #if cls.objtype in AutoDirective._registry:
   #    raise ExtensionError('autodoc documenter for %r is already '
  }}}

**Comment by shimizukawa** [comment]
_From Michael Forbes on 2009-02-19 17:12:23+00:00_

As for the refactoring, it is really great.  The only other issue I found was that it would be nice if there was a special {{{getargspec}}} method that could be overridden.  Instead, I had to do a bit of string fiddling or have a bit of code duplication with an overloaded {{{format_args}}}, but as the whole thing is so small, this is probably not worth doing.

I will attach my minimal zope extension to the previous "enhancement" ticket so people can find it.  It is really nice and short with the new refactoring.  Thanks!

P.S. Is there a better place I should drop the extension as an example or where people can improve it?  Maybe I should just start a small bitbucket project (haven't had time to play with that yet;-)

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2009-02-19 21:57:20+00:00_

Fixed in <<cset 1d128f57fa99de52b0a83e4357bb467cea867803>>. I decided to make _special_attrgetters an attribute of AutoDirective, and add an app.add_autodoc_attrgetter() function to fill it.

The question where to put custom extensions is a good one... for now, it can of course stay on the ticket. A bitbucket projects is always a good idea, too.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2010-08-25 12:19:55+00:00_

Removing milestone: 0.6 (automated comment)
